### PR TITLE
Call actions with default-directory bound to last state dir

### PR DESCRIPTION
**Pull request by clemera**
Before this patch `default-directory` was only bound for calling actions with `ivy-marked-candidates`. This probably was setup to fix the case when the action changes the current directory. 

This patch sets the default-directory for the call of regular and multi-state action, too. This ensures the `default-directory` is always set to the dir of the last state when calling actions.

### CODE REVIEW

**Comment by basil-conto** [comment]
Could you please look into resolving the merge conflicts?

**Comment by clemera** [comment]
Done

File: ivy.el [hunk 1]
```diff
@@ -1496,55 +1496,58 @@ See variable `ivy-recursive-restore' for further information."
...
+                                        (if ivy--directory
+                                            (expand-file-name cand ivy--directory)
+                                          cand)))
+                                    ivy-marked-candidates))
+                                  (multi-action (ivy--get-multi-action ivy-last)))
+                             (if multi-action
+                                 (let ((default-directory (ivy-state-directory ivy-last)))
+                                   (funcall multi-action marked-candidates))
+                               (dolist (c marked-candidates)
+                                 (let ((default-directory (ivy-state-directory ivy-last)))
```
Comment (basil-conto) [inline ivy.el]: Why not bind this only once outside of the `dolist`, or even outside of the surrounding `(if multi-action ...)`?

**Comment by clemera** [comment]
Fixed it, but it's commented out in dir locals:

      ;; (lisp-indent-function . common-lisp-indent-function)

File: ivy.el (hunk 1 above)
Comment (clemera) [reply ivy.el]: Because the actions might change the default directory inside the loop.

Comment (basil-conto) [reply ivy.el]: Ah right, sorry.

**Comment by basil-conto** [comment]
> Fixed it, but it's commented out in dir locals:

Right, which is why it may be inconsistently enforced, but I don't call the shots on that.

**Comment by abo-abo** [comment]
Thanks.

**Comment by basil-conto** [comment]
Note that part of this has been reverted in a877961ecc72753c36654da1c9bf444677f0f02d due to fallout in #2549 and #2610.

**Pull request closed without merging**